Exploration history checkbox

virajprabhu opened the pull request:
Redesigned the exploration history workflow to use a single column of checkboxes instead of two columns of radio buttons. The user is prompted to select exactly two checkboxes to compare the revisions against them. This redesign is in accordance with the discussion on issue thread #801.

seanlip commented:
Thanks very much, @virajprabhu!

@wxyxinyu, please could you take this review? Thanks!

wxyxinyu commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -53,6 +54,32 @@ oppia.controller('ExplorationHistory', [
 
+  $scope.submitVersion = function($event, versionNumber) {
+    var checkbox = $event.target;
+    var id = $scope.versionArr.indexOf(versionNumber);
```
Suggest renaming versionArr to something more descriptive. It's not very clear what this variable is for. Perhaps something like "selectedVersionsArray".

Also it will be good add a comment that this should contain a maximum of two elements.

virajprabhu replied:
Okay, I'll rename and add a comment.

wxyxinyu commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -53,6 +54,32 @@ oppia.controller('ExplorationHistory', [
+      $scope.hideHistoryGraph = true;
+      $scope.hideCompareVersionsButton = false;
+
+      if($scope.versionArr.length == 0)
```
nits: add a space after 'if' and use "===" for comparison.

seanlip replied:
One note (here and below): please always enclose the bodies of if/else blocks with parentheses. This avoids future mistakes like

   if (...)
       statement1;
       statement2;

(where the developer intends statement2 to be executed within the if block, but in practice it actually gets executed all the time).

virajprabhu replied:
Got it, I'll make the necessary changes.

wxyxinyu commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -53,6 +54,32 @@ oppia.controller('ExplorationHistory', [
+
+      if($scope.versionArr.length == 0)
+        $scope.versionCountPrompt = "Please select any two.";
+      else if($scope.versionArr.length == 1)
```
same

wxyxinyu commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -53,6 +54,32 @@ oppia.controller('ExplorationHistory', [
+      else 
+        $scope.versionCountPrompt = "Please select exactly two.";      
+    }
+  }
```
End function declaration with a semicolon.

wxyxinyu commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -53,6 +54,32 @@ oppia.controller('ExplorationHistory', [
+      else if($scope.versionArr.length == 1)
+        $scope.versionCountPrompt = "Please select one more.";
+      else 
+        $scope.versionCountPrompt = "Please select exactly two.";      
```
This seems to differ from the mock that @ben-targan gave. Ben's mock had checkboxes disabled after 2 were checked, while this prompts the user to select less. Did you discuss this with @ben-targan?

seanlip replied:
I think it would be better to follow the flow in Ben's mock -- i.e. disable the checkboxes. @virajprabhu, would it be possible to do this?

virajprabhu replied:
Sure, I'll disable the checkboxes after two are selected.

virajprabhu replied:
Hi @seanlip, my understanding of disabling the checkboxes is as follows: after two are selected,  the <i>remaining</i> checkboxes should be disabled, since the user should still be able to deselect one or both of the currently selected ones(at which point the others will be enabled as well). Is this what you mean?

seanlip replied:
Yes, that's right. Thanks!

wxyxinyu commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -120,112 +150,108 @@ oppia.controller('ExplorationHistory', [
   // Function to set compared version metadata, download YAML and generate
   // diff graph and legend when selection is changed
-  $scope.changeCompareVersion = function(versionNumber, changedVersion) {
+  $scope.changeCompareVersion = function(versionNumber) {
```
Is it actually necessary for this function to take versionNumber as a parameter?

virajprabhu replied:
It wasn't being used in the original version either, so I didn't think to change it. Will remove it now.

wxyxinyu commented on core/templates/dev/head/editor/exploration_history.html:
```diff
@@ -18,8 +18,7 @@
       <div class="col-sm-2 col-md-2 col-lg-2">
-        <input type="radio" name="compareVer1" ng-model="compareVersions.selectedVersion1" ng-value="versionNumber" ng-change="changeCompareVersion(versionNumber, 'selectedVersion1')" class='protractor-test-history-v1-selector'>
-        <input type="radio" name="compareVer2" ng-model="compareVersions.selectedVersion2" ng-value="versionNumber" ng-change="changeCompareVersion(versionNumber, 'selectedVersion2')" class='protractor-test-history-v2-selector'>
+        <input type="checkbox" name="compareVer1" ng-model="versionArr" ng-click="submitVersion($event, versionNumber)" class='protractor-test-history-v1-selector'>
```
It doesn't seem to make sense to name this "compareVer1" when there's only 1 column of checkboxes now.

The ng-model tag is unnecessary.

wxyxinyu replied:
The 'protractor-test-history-v1-selector' (and the v2-selector in the old version) is used in integration tests. (See tests/protractor/editorAndPlayer.js and tests/protractor_utils/editor.js for the code, and https://github.com/oppia/oppia/wiki/Writing-Integration-Tests for an overview of the tests.) The integration tests also need to be changed to use checkboxes instead of radio buttons. (The relevant code is near the end of tests/protractor_utils/editor.js)

virajprabhu replied:
Thanks, I'll read up on the tests section and make the required change in the integration tests.

wxyxinyu commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -53,6 +54,32 @@ oppia.controller('ExplorationHistory', [
     }
   });
 
+  $scope.submitVersion = function($event, versionNumber) {
```
Suggest renaming this function something like "changeSelectedVersions". This is called when checked checkboxes are changed, which seems to be a different action from submitting versions to compare (when the "Compare Selected Versions" button is clicked).

I'm not entirely sure about this though, since the versions are actually sent to the comparison function in here. Maybe it will make more sense to submit both versions when the button is clicked? (Originally, versions were submitted as the radio buttons were changed so that users could change the compared versions without clicking the button again, but now this is no longer possible.)

virajprabhu replied:
I'll make a change to compare versions only when the button is clicked.

wxyxinyu commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -53,6 +54,32 @@ oppia.controller('ExplorationHistory', [
+      $scope.versionArr.splice(id, 1);
+    }
+    
+    if($scope.versionArr.length == 2) {
```
nit: add space before the '(' here too.

virajprabhu commented:
@seanlip , @wxyxinyu , I've made a commit with the following changes:
- The compare button and prompt are disabled if there are less than two version in total
- The checkboxes are disabled on selecting two
- The function to compare is only called on clicking the button
- The variables have been renamed as suggested

The following changes remain:
- Changes to the integration tests
- Presently, the checkboxes are not cleared when the page is reloaded, though the variables are. This can lead to some confusing behavior. I'm thinking of clearing the checkboxes inside the refreshVersionHistory() function. Does this sound okay?

I'll try to complete the remaining changes by tomorrow. Please have a look and tell me if the current changes are okay. Thanks!

wxyxinyu commented:
This is very nice, thanks!

Nice use of comments throughout. 

As well as hiding the compare button and prompt, I think the checkbox should also be hidden when there is only 1 version?

wxyxinyu commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -70,9 +107,18 @@ oppia.controller('ExplorationHistory', [
-      $scope.hideCompareVersionsButton = false;
+
+      // Disable all comparisons if there are less than two revisions in total.
+      $scope.disableCompare = (currentVersion<2) ? true : false;
```
nit: add spaces before and after the operator
This line can be simplified as
    $scope.disableCompare = (currentVersion < 2);

virajprabhu replied:
I've made this change.

wxyxinyu commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -252,11 +295,15 @@ oppia.controller('ExplorationHistory', [
 
+  $scope.isCheckboxDisabled = function(versionNumber) {
+    if ($scope.selectedVersionsArray.length === 2) {
+      return !($scope.selectedVersionsArray.indexOf(versionNumber) > -1);
```
I think it's clearer to write
    return $scope.selectedVersionsArray.indexOf(versionNumber) == -1;
here.

virajprabhu replied:
Done.

wxyxinyu commented:
And, yes I think clearing the checkboxes in refreshVersionHistory() is a good idea.

seanlip commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -53,6 +54,42 @@ oppia.controller('ExplorationHistory', [
+  // Compares the two selected versions and displays the comparison results.
+  $scope.compareSelectedVersions = function() {
+    if($scope.selectedVersionsArray.length === 2) {
+      console.log('comparing selected versions');
```
Please remove console.logs before submitting.

virajprabhu replied:
Done.

seanlip commented:
Hi @virajprabhu, thanks again for working on this!

Just a quick note about the integration tests: if you're having trouble running them, it may be necessary to replace the chromedriver binary in ../node_modules/protractor/selenium, per [this thread](https://groups.google.com/forum/#!topic/oppia-dev/5lb46KU8_30).

seanlip commented:
Update: I just updated the integration tests version in develop, so if you merge that into this branch, they should work fine!

virajprabhu commented:
Thanks! I'm away for the weekend so I'll make the remaining changes in a
day or two if that's alright
On Oct 24, 2015 5:41 AM, "Sean Lip" <notifications@github.com> wrote:

> Update: I just updated the integration tests version in develop, so if you
> merge that into this branch, they should work fine!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/oppia/oppia/pull/1039#issuecomment-150725774>.
>

wxyxinyu commented:
No problem, thanks for working on this!

virajprabhu commented:
Hi @wxyxinyu, I've made most changes. I had a question regarding clearing checkboxes in refreshVersionHistory() - if the user has selected some checkboxes and then goes on to a different tab, (say to edit the exploration), and comes back to the history tab - is it okay to have the checkboxes still selected?

seanlip commented:
Hi @virajprabhu, thanks for making the changes! One quick note: would you mind replying to each comment, following the instructions in step 5 [here](https://github.com/oppia/oppia/blob/develop/CONTRIBUTING.md#instructions-for-making-a-code-change)? This helps the reviewer(s) keep track of which comments have been addressed, and which haven't. You can just say "Done." for any comments that have been fully addressed.

The question you raise is a good one. I would lean towards clearing them when the history tab is loaded, since the different tabs 'feel' like different pages.

virajprabhu commented:
Hi @seanlip , @wxyxinyu. I've pushed a commit with all but two of the suggested changes:
- Clearing check boxes when the history tab is navigated to: Clearing them in refreshVersionHistory() didn't achieve the desired goal, so I'm unsure of the best way to go about this. Any suggestions?
- Integration tests: I'm working on this right now.

seanlip commented:
Hm. I believe the 'refreshVersionHistory' event is triggered on a tab switch to the history page. Maybe this needs a bit of debugging: could you check whether the $scope.$on('refreshVersionHistory', ...) is indeed called? And, if so, could you check whether the inner $scope.refreshVersionHistory() is also called? You can add temporary console.logs() to debug.

seanlip commented:
It may be that you need to add the user-option-resetting logic to the $scope.$on rather than to $scope.refreshVersionHistory, since the latter is (I think) more for when the history data itself changes.

wxyxinyu commented:
This is strange -- the travis tests seemed to be failing last night but pass now?

Also, @virajprabhu if you need any help with the integration tests don't hesitate to ask!

seanlip commented:
The travis thing is my doing -- the failure was due to a temporary codecov issue, I think, unrelated to this commit, so I restarted the job.

virajprabhu commented:
Hi @seanlip, I've pushed a fix that clears the checkboxes when the history tab is loaded.  @wxyxinyu, I didn't actually get time to read up on the tests. I'll do so now, and let you know if I need any help. Thanks!

virajprabhu commented:
Hi @wxyxinyu, I wanted to know what's the recommended way to debug integration tests? Do I have to run all the tests every time?

wxyxinyu commented:
To run just one test, change "it('should display the history' ..." to "fit('should display the history'...". You can also use 'ddescribe' in place of 'describe' to run just one group of tests (though in this case there is only 1 test in the group).

seanlip commented:
Just FYI, I think now it's `fdescribe` instead of `ddescribe` (at least since we upgraded Protractor). I recently updated the docs [here](https://github.com/oppia/oppia/wiki/Running-Tests#other-notes).

virajprabhu commented:
Hi @wxyxinyu, @seanlip, I went through the protractor docs and tried to rewrite the tests inside _selectComparedVersions to use checkboxes. I tried using css selectors to only get elements of the protractor-test-history-v1-selector class (which should be both checkboxes), but I'm getting an click undefined on element error. I'm not sure what the v1 and v2 that the historyTab protractor test passes to this function through a wrapper are for. Could you help me out here?

wxyxinyu commented:
v1 and v2 are the two versions numbers to be compared. 

Your method sounds right, are you doing something like "element.all(by.css('.protractor-test-history-v1-selector')).get(position).click();"? Could you try sending me your modified files in a gist so I can try to see what's going on? Thanks.

seanlip commented:
Hi @virajprabhu -- just checking in; how's the PR going?

Also, just FYI, we have a chat room [here](https://gitter.im/oppia), in case you'd like to talk through anything!

virajprabhu commented:
Hi @seanlip, @wxyxinyu . Sorry about the delay, had a busy week at work. I haven't been able to get the protractor test for the checkbox running correctly. Particularly I'm confused about this portion of the original code of `_selectComparedVersions()` in `protractor_tests/editor.js`:

    element.all(by.css('.protractor-test-history-v1-selector')).first()
        .getAttribute('value').then(function(versionNumber) {
        v1Position = versionNumber - v1;
        v2Position = versionNumber - v2;

First off, I'll have to modify the 'value' attribute since for a checkbox this will equal 'on' or 'off'. But my doubt is this:

I see that first() will return the highest version number on the page, since they are arranged in descending order. So if we want to compare versions 1 and 2, we pass v1=1, v2=2  when there are say, 5 versions (and hence `versionNumber=4`). The code will effectively `get(3)` and `get(2)` which will be revisions 3 and 2, respectively. Isn't this off by one (since we wanted to compare 1 and 2)? And is my understanding of the subtraction correct? Thanks.

wxyxinyu commented:
The version numbers are 1-indexed, so the version number of the top selector will be 5 in your case. 

After that, the code will call get(4) and get(3). get(n) finds the nth element with the css class on the page, counting from the top of the page. 

Does this help?

virajprabhu commented:
Hi oppia/oppia,

I'd like to add you to my professional network on LinkedIn.

- Viraj

Confirm that you know Viraj: https://www.linkedin.com/e/v2?e=0-igz4yil1-20&t=ssuw&tracking=eml-guest-invite-cta&ek=invite_guest&sharedKey=x9tN1R27&invitationId=6071288953372241920








You received an invitation to connect. LinkedIn will use your email address to make suggestions to our members in features like People You May Know. Unsubscribe here: https://www.linkedin.com/e/v2?e=0-igz4yil1-20&t=lun&midToken=AQFmrFVhGEd36Q&ek=invite_guest&loid=AQGWdo92SP6J9AAAAVEGNR4SoHVHfw6Wjl6nJrQoMdQnbD47D2fabHVAdQiDmcSFfMz_F2fOAkcqrwQmxnhxC9zwBTemcCEBeZcLRlq1dLU239AEBdxhlVT4ndNI50o-Sl0YQ_lSFFu1oI-qWMS5-FZB0stzBJ5BFBN09c5R_-tm-L6eU7BJfcCxNhOVp3v1nVMw&eid=0-igz4yil1-20


This email was sent to reply@reply.github.com.

If you need assistance or have questions, please contact LinkedIn Customer Service: https://www.linkedin.com/e/v2?e=0-igz4yil1-20&a=customerServiceUrl&ek=invite_guest

&copy; 2015 LinkedIn Corporation, 2029 Stierlin Court, Mountain View CA 94043. LinkedIn and the LinkedIn logo are registered trademarks of LinkedIn.

virajprabhu commented:
Hi @wxyxinyu, sorry for the delayed response. That did help. I've not been able to find a replacement for this line yet:
    element.all(by.css('.protractor-test-history-v1-selector')).first()
    .getAttribute('value')
I'm not familiar with the `codemirrorMergeview` directive that extends the element to set the value attribute. Should the watch still be set on `attrs.leftValue` to work with the checkbox?

wxyxinyu commented:
Hey @virajprabhu, 
That line is just meant to get the largest version number on the page. It finds the first element on the page with a 'protractor-test-history-v1-selector' class, and finds its 'value' attribute, which is the version number associated with that row. 

In the codemirrorMergeview, the attrs.leftValue is the YAML string that populates the codemirror view on the left. This shouldn't have anything to do with what you're trying to do now.

Also, if you like, we can chat in real time about this, maybe on the gitter chatroom at gitter.im/oppia. I think this would be really helpful to understand the protractor tests and clear up any remaining doubts.

virajprabhu commented:
Sure @wxyxinyu, when would it be a good time to have a chat on gitter?

Hey @virajprabhu <https://github.com/virajprabhu>,
That line is just meant to get the largest version number on the page. It
finds the first element on the page with a
'protractor-test-history-v1-selector' class, and finds its 'value'
attribute, which is the version number associated with that row.

In the codemirrorMergeview, the attrs.leftValue is the YAML string that
populates the codemirror view on the left. This shouldn't have anything to
do with what you're trying to do now.

Also, if you like, we can chat in real time about this, maybe on the gitter
chatroom at gitter.im/oppia. I think this would be really helpful to
understand the protractor tests and clear up any remaining doubts.

—
Reply to this email directly or view it on GitHub
<https://github.com/oppia/oppia/pull/1039#issuecomment-156869512>.

wxyxinyu commented:
I'm free tomorrow (16 Nov) after 2.30pm EST (till 12-ish), or the day after after 4.30pm EST. (Sorry, I don't know what timezone you're in, I could try to find another time if these are inconvenient.)

virajprabhu commented:
Hi @wxyxinyu, around 2.30 pm EST works for me. Will that be possible? Just FYI, I follow Indian Standard Time.

wxyxinyu commented:
@virajprabhu Sure, see you then.

virajprabhu commented:
Hi @wxyxinyu, @seanlip. The protractor test visually seems to be doing the right thing now. However, it later crashes with the message `element is not clickable at point`. I read somewhere that this could be an issue with the Selenium WebDriver. Any ideas on how I could fix this?

seanlip commented:
There'll usually be an error traceback in the terminal. Can you paste it here?

Another thing I usually do is watch the test running (by eye) and look for the point in the test at which it fails -- this usually sheds light on why it's failing.

virajprabhu commented:
Hi @seanlip, @wxyxinyu: I was able to fix the issue with the test and it works fine now. I think I've addressed all the review comments so far, could you take a look and let me know if any more changes are required? Thanks!

seanlip commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -48,20 +49,61 @@ oppia.controller('ExplorationHistory', [
   var nodesData = null;
 
   $scope.$on('refreshVersionHistory', function(evt, data) {
+      $scope.clearAllCheckboxes();
```
indentation wrong -- please deindent by 2

virajprabhu replied:
Done.

seanlip commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -48,20 +49,61 @@ oppia.controller('ExplorationHistory', [
+  
+  // Compares the two selected versions and displays the comparison results.
+  $scope.compareSelectedVersions = function() {
+    if($scope.selectedVersionsArray.length === 2) {
```
nit: single space after if

virajprabhu replied:
Done.

seanlip commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -48,20 +49,61 @@ oppia.controller('ExplorationHistory', [
+    }
+  };
+  // Changes the checkbox selection and provides an appropriate user prompt.
+  $scope.changeSelectedVersions = function($event, versionNumber) {
```
Let's use 'evt' as the first variable (this is consistent with other parts of the codebase which do similar things).

virajprabhu replied:
Done.

seanlip commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -48,20 +49,61 @@ oppia.controller('ExplorationHistory', [
+  // Changes the checkbox selection and provides an appropriate user prompt.
+  $scope.changeSelectedVersions = function($event, versionNumber) {
+    var checkbox = $event.target;
+    var id = $scope.selectedVersionsArray.indexOf(versionNumber);
```
Can this have a more descriptive name?

virajprabhu replied:
Are you referring to the 'id' variable?

virajprabhu replied:
Changed `id` to `selectedVersionsArrayPos`.

seanlip commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -48,20 +49,61 @@ oppia.controller('ExplorationHistory', [
+    
+    if ($scope.selectedVersionsArray.length === 2) {
+      $scope.versionCountPrompt = "";
+      // Disable checkboxes
```
What does this comment refer to?

virajprabhu replied:
It was out of place. Fixed now.

seanlip commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -48,20 +49,61 @@ oppia.controller('ExplorationHistory', [
+    }
+    
+    if ($scope.selectedVersionsArray.length === 2) {
+      $scope.versionCountPrompt = "";
```
Here and elsewhere: we use single quotes for strings in JS

virajprabhu replied:
Done.

seanlip commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -70,16 +112,25 @@ oppia.controller('ExplorationHistory', [
+      // Disable all comparisons if there are less than two revisions in total.
+      $scope.disableCompare = (currentVersion < 2);
+
+      $scope.hideCompareVersionsButton = $scope.disableCompare;
```
The name reads like a function that does something, since it starts with a verb. Maybe call it $scope.compareVersionsButtonIsHidden?

virajprabhu replied:
Done.

seanlip commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -70,16 +112,25 @@ oppia.controller('ExplorationHistory', [
-      $scope.hideCompareVersionsButton = false;
+
+      // Disable all comparisons if there are less than two revisions in total.
+      $scope.disableCompare = (currentVersion < 2);
```
The name reads like a function that does something, since it starts with a verb. Maybe call it $scope.comparisonsAreDisabled?

virajprabhu replied:
Done.

seanlip commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -70,16 +112,25 @@ oppia.controller('ExplorationHistory', [
 
         // Re-populate snapshotOrderArray and explorationVersionMetadata when
-        // history is refreshed.
+        // history    is refreshed.
```
superfluous extra spaces

virajprabhu replied:
Fixed.

seanlip commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -90,11 +141,11 @@ oppia.controller('ExplorationHistory', [
             'versionNumber': explorationSnapshots[i].version_number
           };
-          $scope.snapshotOrderArray.push(explorationSnapshots[i].version_number);
+          $scope.snapshotOrderArray.push({vnum:explorationSnapshots[i].version_number, selected:false});
```
can this be broken up a bit?

```
    $scope.snapshotOrderArray.push({
      vnum: explorationSnapshots[i].version_number,
      selected: false
    });
```

virajprabhu replied:
Done.

wxyxinyu replied:
I would suggest using versionNumber instead of vnum, and renaming snapshotOrderArray to versionCheckboxArray (since it no longer just stores the order of displayed snapshots).

seanlip commented on core/templates/dev/head/editor/ExplorationHistory.js:
```diff
@@ -252,11 +299,16 @@ oppia.controller('ExplorationHistory', [
-      $scope.compareVersions &&
-      $scope.compareVersions.hasOwnProperty('selectedVersion1') &&
-      $scope.compareVersions.hasOwnProperty('selectedVersion2'));
+      $scope.compareVersions && ($scope.selectedVersionsArray.length === 2));
```
probably don't need the inner brackets

virajprabhu replied:
Done.

seanlip commented on core/templates/dev/head/editor/exploration_history.html:
```diff
@@ -18,24 +18,23 @@
     <h3>List of Changes</h3>
     <div ng-repeat="versionNumber in snapshotOrderArray" class="row">
```
Should we still be calling this 'versionNumber'? It seems like that would be a number, but now it's an object.

Maybe snapshotMetadata or something like that?

wxyxinyu replied:
I think it is now an object representing a checkbox, so I would suggest calling it versionCheckbox and versionCheckboxArray

virajprabhu replied:
Done.

seanlip commented on core/templates/dev/head/editor/exploration_history.html:
```diff
@@ -44,12 +43,16 @@
 
   <md-card class="oppia-editor-card" ng-hide="hideHistoryGraph || !explorationVersionMetadata" style="position: relative;">
-    <h3>Changes from version <[compareVersionMetadata.earlierVersion.versionNumber]> to version <[compareVersionMetadata.laterVersion.versionNumber]></h3>
+    <h3>Changes from version <[compareVersionMetadata.earlierVersion.versionNumber.vnum]> to version <[compareVersionMetadata.laterVersion.versionNumber]></h3>
```
Should the laterVersion also have a vnum reference?

virajprabhu replied:
That was a mistake, fixed now.

seanlip commented:
Hi @virajprabhu, thanks a lot! I took a quick, surface-level pass (mainly looking at style, consistency, naming, etc.) and had some comments.

@wxyxinyu, would you mind doing a more thorough review (for the logic/UI)?

wxyxinyu commented on core/tests/protractor_utils/editor.js:
```diff
@@ -974,13 +974,32 @@ var _runFromHistoryTab = function(callbackFunction) {
+  element.all(by.css('.protractor-test-history-checkbox-selector')).count()
+      .then(function(versionNumber) {
+    if (v1 < 0) {
+      throw Error('In editor._selectComparedVersions(' + v1 + '),' +
```
shouldn't this be 
    throw Error('In editor._selectComparedVersions(' + v1 + ', ' + v2 + '),' +

ditto for the ones below.

virajprabhu replied:
Done.

virajprabhu commented:
Hi @seanlip, @wxyxinyu. I've pushed the latest changes. Thanks.

wxyxinyu commented:
This looks great. I think it's ready to merge. @virajprabhu, could you merge develop into this branch?
